agent/agent_test.go

```diff
@@ -14,11 +20,26 @@ import (
 	"github.com/docker/swarmkit/connectionbroker"
 	"github.com/docker/swarmkit/remotes"
 	"github.com/docker/swarmkit/testutils"
+	"github.com/docker/swarmkit/xnet"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/net/context"
 )
 
+var localDispatcher = false
+
+// TestMain runs every test in this file twice - once with a local dispatcher, and
```

Comment:
> Are there separate code paths for local and remote dispatchers?

Sort of, although it's not here.  We are adding the extra context so we can cancel it because closing the connection doesn't work on local connections.  `connectionbroker` ignores closes on local connections, so the session can't actually be restarted if it's a local connection, if we just close the connection.  This is sort of a regression test - it fails without the context changes in the rest of this PR.